### PR TITLE
vine: daskvine scheduling algorithm

### PR DESCRIPTION
**Pull request by JinZhou5042**
## Proposed Changes

The previous PR #3923 was closed by accident ... 

## Merge Checklist

The following items must be completed before PRs can be merged.
Check these off to verify you have completed all steps.

- [ ] `make test`       Run local tests prior to pushing.
- [ ] `make format`     Format source code to comply with lint policies. Note that some lint errors can only be resolved manually (e.g., Python)
- [ ] `make lint`       Run lint on source code prior to pushing.
- [ ] Manual Update:     Update the manual to reflect user-visible changes.
- [ ] Type Labels:       Select a github label for the type: bugfix, enhancement, etc.
- [ ] Product Labels:    Select a github label for the product: TaskVine, Makeflow, etc.
- [ ] PR RTM:            Mark your PR as ready to merge.

### CODE REVIEW

File: taskvine/src/bindings/python3/ndcctools/taskvine/task.py [hunk 1]
```diff
@@ -829,6 +829,13 @@ def resources_allocated(self):
             return None
         return cvine.vine_task_get_resources(self._task, "allocated")
 
+    ##
+    # Get the execution time of this task in seconds.
+    #
+    @property
+    def execution_time(self):
```
Comment (btovar) [inline taskvine/src/bindings/python3/ndcctools/taskvine/task.py:836]: I think this is already available with: `t.resources_measured.wall_time` ?

File: taskvine/src/manager/taskvine.h [hunk 2]
```diff
@@ -452,9 +452,15 @@ regardless of the priority.
...
 */
-
 void vine_task_set_priority(struct vine_task *t, double priority);
 
+/** Get the actual execution time of the task.
+execution_time = t->time_workers_execute_last_end - t->time_workers_execute_last_start.
+@param t A task object.
+@return The actual execution time of the task in seconds.
+*/
+double vine_task_get_execution_time(struct vine_task *t);
```
Comment (btovar) [inline taskvine/src/manager/taskvine.h:462]: I don't think this function is needed (see above).

**Review by btovar — CHANGES_REQUESTED** [review]

File: taskvine/src/bindings/python3/ndcctools/taskvine/dask_executor.py [hunk 3]
```diff
@@ -273,6 +281,7 @@ def _dask_execute(self, dsk, keys):
                         print(f"{t.key} ran on {t.hostname}")
 
                     if t.successful():
+                        self.category_execution_time[t.category].append(t.execution_time)
```
Comment (btovar) [inline taskvine/src/bindings/python3/ndcctools/taskvine/dask_executor.py:284]: I don't think you want to keep all the execution times around in that list, as it looks like a potential memory leak?
Probably you can do a rolling average (a variable to keep the sum of the execution times and a variable to keep how many tasks of that category.)

**Review by btovar — CHANGES_REQUESTED** [review]

File: taskvine/src/bindings/python3/ndcctools/taskvine/dask_executor.py [hunk 4]
```diff
@@ -281,7 +281,9 @@ def _dask_execute(self, dsk, keys):
                         print(f"{t.key} ran on {t.hostname}")
 
                     if t.successful():
-                        self.category_execution_time[t.category].append(t.execution_time)
+                        cat_total_time = self.category_execution_time[t.category]["avg_execution_time"] * self.category_execution_time[t.category]["num_tasks"]
+                        self.category_execution_time[t.category]["num_tasks"] += 1
+                        self.category_execution_time[t.category]["avg_execution_time"] = (cat_total_time + t.resources_measured.wall_time) / self.category_execution_time[t.category]["num_tasks"]
```
Comment (btovar) [inline taskvine/src/bindings/python3/ndcctools/taskvine/dask_executor.py:286]: You can simply do `+= t.resources_measured.wall_time` without the need of `cat_total_time`.

File: taskvine/src/bindings/python3/ndcctools/taskvine/dask_executor.py [hunk 5]
```diff
@@ -175,7 +175,7 @@ def get(self, dsk, keys, *,
             self.wrapper = wrapper
             self.wrapper_proc = wrapper_proc
             self.prune_files = prune_files
-            self.category_execution_time = defaultdict(list)
+            self.category_execution_time = defaultdict(lambda: {"num_tasks": 0, "avg_execution_time": 0})
```
Comment (btovar) [inline taskvine/src/bindings/python3/ndcctools/taskvine/dask_executor.py:178]: Remove the "avg_", as that would be confusing. You are keeping the total execution time, not the average.

File: taskvine/src/bindings/python3/ndcctools/taskvine/dask_executor.py (hunk 4 above)
Comment (btovar) [inline taskvine/src/bindings/python3/ndcctools/taskvine/dask_executor.py:286]: It is easier and more numerically precise to simply keep the total execution time and compute the average only when needed.

**Review by btovar — CHANGES_REQUESTED** [review]

File: taskvine/src/bindings/python3/ndcctools/taskvine/dask_executor.py (hunk 4 above)
Comment (JinZhou5042) [reply taskvine/src/bindings/python3/ndcctools/taskvine/dask_executor.py]: @btovar Thanks! That's a very good point!

**Review by btovar — APPROVED** [review]

**Pull request merged**